lib/Sema/CSDiag.cpp

```diff
@@ -3659,10 +3684,33 @@ bool FailureDiagnosis::visitApplyExpr(ApplyExpr *callExpr) {
     // Re-type-check the argument with the expected type of the candidate set.
     // This should produce a specific and tailored diagnostic saying that the
     // type mismatches with expectations.
-    if (!typeCheckChildIndependently(badArgExpr,
-                                     calleeInfo.failedArgument.parameterType,
+    Type paramType = calleeInfo.failedArgument.parameterType;
+    if (!typeCheckChildIndependently(badArgExpr, paramType,
                                      CTP_CallArgument, options))
       return true;
+      
+    // If that fails, it could be that the argument doesn't conform to an archetype.
+    Type argType = badArgExpr->getType();
+
+    if (paramType->is<ArchetypeType>() && !argType->hasTypeVariable() &&
```

Comment:
As I think about this more, I don't think this diagnosis can apply to subscript expressions. The parser doesn't allow generic parameters on subscripts (is that intended? or just a current limitation?). Is there some other way to get subscripts with unresolved archetypes that I'm not thinking of?

On the other hand, the diagnosis of unresolved members expressions could be improved by this:
```
struct A {
  //subscript<T: Comparable>(x: T) { }
  func foo<T: Comparable>(x: T) { }
}
struct B {}
let _ = A().foo(B())
```
Would you prefer I go ahead and work on that here, or in a separate pull after this lands?